lib/AST/ArchetypeBuilder.cpp

```diff
@@ -1030,6 +1036,37 @@ auto ArchetypeBuilder::resolveArchetype(Type type, PotentialArchetype *basePA)
   return nullptr;
 }
 
+auto ArchetypeBuilder::FIXMEresolveArchetypeOrTypeThroughTypeAlias(
+    Type type, PotentialArchetype *basePA)
+    -> llvm::PointerUnion<PotentialArchetype *, Type> {
+  auto pa = resolveArchetype(type, basePA);
+  if (!pa)
+    return type;
+
+  pa = pa->getRepresentative();
+  if (!pa->getParent() || !pa->getTypeAliasDecl())
+    return pa;
+
+  assert(llvm::all_of(pa->getEquivalenceClass(),
+                      [&](PotentialArchetype *pa) {
+                        return pa->getParent() && pa->getTypeAliasDecl();
+                      }) &&
+         "unexpected typealias representative with non-typealias equivalent");
+
+  // The right-hand side of the typealias could itself be an archetype
+  // (e.g. protocol P { associatedtype A; typealias B = A }), so we need to
+  // resolve that.
+  auto concrete = pa->getConcreteType();
+  auto rhsPA = resolveArchetype(concrete, basePA);
```

Comment:
The assertion above is the reason we don't need to recursively call `resolveArchetypeOrTypeThroughTypeAlias` here, right?